File: src/utils/wind.ts
import { rad, deg } from './navigationMath';

/**
 * 真風向・風速 (True Wind)
 * 相対風向(船首基準)、相対風速、船の針路、船速 → 真風向、真風速
 */
export function trueWind(
  relativeDir: number,
  relativeSpeed: number,
  shipCourse: number,
  shipSpeed: number,
) {
  // 相対風のベクトル（船体座標→地球座標）
  const apparentDir = shipCourse + relativeDir;
  const appR = rad(apparentDir);
  const scR = rad(shipCourse);

  // Apparent wind vector (where wind comes FROM)
  const awx = relativeSpeed * Math.sin(appR);
  const awy = relativeSpeed * Math.cos(appR);

  // Ship motion vector
  const sx = shipSpeed * Math.sin(scR);
  const sy = shipSpeed * Math.cos(scR);

  // True wind = apparent wind - ship motion (wind FROM direction)
  const twx = awx - sx;
  const twy = awy - sy;

  const trueSpeed = Math.sqrt(twx * twx + twy * twy);
  let trueDir = deg(Math.atan2(twx, twy));
  if (trueDir < 0) trueDir += 360;

  return { trueDir, trueSpeed };
}

File: NavCalc/NavCalc.Core/Calculations/Wind.cs
using System;
using static NavCalc.Core.Calculations.NavigationMath;

namespace NavCalc.Core.Calculations
{
    public static class Wind
    {
        /// <summary>
        /// 真風向・風速 (True Wind)
        /// </summary>
        public static (double TrueDir, double TrueSpeed) TrueWind(
            double relativeDir, double relativeSpeed, double shipCourse, double shipSpeed)
        {
            double apparentDir = shipCourse + relativeDir;
            double appR = Rad(apparentDir);
            double scR = Rad(shipCourse);

            double awx = relativeSpeed * Math.Sin(appR);
            double awy = relativeSpeed * Math.Cos(appR);

            double sx = shipSpeed * Math.Sin(scR);
            double sy = shipSpeed * Math.Cos(scR);

            double twx = awx - sx;
            double twy = awy - sy;

            double trueSpeed = Math.Sqrt(twx * twx + twy * twy);
            double trueDir = Deg(Math.Atan2(twx, twy));
            if (trueDir < 0) trueDir += 360.0;

            return (trueDir, trueSpeed);
        }
    }
}
